optuna/terminator/improvement/evaluator.py

```diff
@@ -128,3 +128,53 @@ def _validate_input(
                 "The intersection search space is empty. This condition is not supported by "
                 f"{cls.__name__}."
             )
+
+
+@experimental_class("3.2.0")
+class BestValueStagnationEvaluator(BaseImprovementEvaluator):
+    """
+    Evaluates the stagnation period of the best value in an optimization process.
+
+    This class is initialized with a maximum stagnation period (`max_stagnation_trials`)
+    and is designed toevaluate the remaining trials before reaching this maximum period
+    of allowed stagnation.
+
+    Args:
+        - max_stagnation_trials (int): The maximum number of trials allowed for stagnation.
+
+    Example Usage:
+        >>> evaluator = BestValueStagnationEvaluator(max_stagnation_trials=30)
+        >>> stagnated_trials = evaluator.evaluate()
+    """
+
+    def __init__(
+        self,
+        max_stagnation_trials: int = 30,
+    ):
+        if max_stagnation_trials < 0:
+            raise ValueError("The number of tolerance of step must not be negative.")
+        self._max_stagnation_trials = max_stagnation_trials
+
+    def evaluate(
+        self,
+        trials: List[FrozenTrial],
+        study_direction: StudyDirection,
+    ) -> int:
+        is_maximize_direction = True if (study_direction == StudyDirection.MAXIMIZE) else False
+        trials = [trial for trial in trials if trial.state == TrialState.COMPLETE]
+        if len(trials) == 0:
+            return self._max_stagnation_trials
+        current_step = len(trials) - 1
+
+        best_step = 0
+        for i_step, trial in enumerate(trials):
+            best_value = trials[best_step].value
+            i_step_value = trial.value
+            assert best_value is not None
+            assert i_step_value is not None
+            if is_maximize_direction and (best_value < i_step_value):
+                best_step = i_step
+            elif (not is_maximize_direction) and (best_value > i_step_value):
+                best_step = i_step
```

Comment:
As `i_step` is not referred to outside of this loop, changing it to a more generic name like `i` would likely make the code more readable.
Similarly, renaming `i_step_value` to `current_value` would make the contrast with `best_value` clearer.

```suggestion
        for i, trial in enumerate(trials):
            best_value = trials[best_step].value
            current_value = trial.value
            assert best_value is not None
            assert current_value is not None
            if is_maximize_direction and (best_value < current_value):
                best_step = i
            elif (not is_maximize_direction) and (best_value > i_step_value):
                best_step = i
```